lib/network_monitor.c

```diff
@@ -53,37 +53,64 @@ struct mptcpd_nm
 //               Network Address Information Handling
 // -------------------------------------------------------------------
 
+/**
+ * @struct mptcpd_rtm_addr
+ *
+ * @brief Encapsulate network address information.
+ *
+ * This is a convenience structure that encapsulates rtnetlink
+ * @c RTM_NEWADDR, @c RTM_DELADDR, and @c RTM_GETADDR message data for
+ * a given network address.
+ */
+struct mptcpd_rtm_addr
+{
+        /**
+         * @c RTM_NEWADDR, @c RTM_DELADDR, and @c RTM_GETADDR message
+         * data.
+         */
+        struct ifaddrmsg const *ifa;
+
+        /**
+         * Network address retrieved from @c IFA_ADDRESS rtnetlink
+         * attribute.
+         */
+        void const *const addr;
+};
+
 /**
  * @brief Create an object that contains internet network
  *        address-specific information.
  *
- * @param[in] ifa Network address-specific information retrieved
- *                from the @c RTM_NEWADDR message.
- * @param[in] len Length of the @C RTM_NEWADDR Netlink message,
- *                potentially including @c rtattr attributes.
+ * @param[in] info Network address-specific information retrieved from
+ *                 the @c RTM_NEWADDR message.
  */
 static struct mptcpd_in_addr *
-mptcpd_in_addr_create(struct ifaddrmsg const *ifa, void const *data)
+mptcpd_in_addr_create(struct mptcpd_rtm_addr const *info)
 {
-        assert(ifa != NULL);
+        assert(info != NULL);
+
+        sa_family_t const family = info->ifa->ifa_family;
+
+        if (family != AF_INET && family != AF_INET6)
+                return NULL;
 
         struct mptcpd_in_addr *const address =
                 l_new(struct mptcpd_in_addr, 1);
 
-        address->family = ifa->ifa_family;
+        address->family = family;
 
-        if (ifa->ifa_family == AF_INET) {
+        if (family == AF_INET) {
                 /*
                   Kernel nla_put_in_addr() inserts a big endian 32
                   bit unsigned integer, not struct in_addr.
                 */
                 in_addr_t const sa =
-                        *(uint32_t const*) data;
+                        *(uint32_t const*) info->addr;
```

Comment:
Do you need the intermediate `sa` variable here?

The `in_addr_t` vs `uint32_t` mismatch on each side of the assignment catches my eye. I realize they're equivalent, does it help anything to use them this way?

Reply:
Not at all.  I'll drop the in_addr_t in this case, and the other one you mentioned.